pkg/trace/writer/trace.go

```diff
@@ -139,6 +143,37 @@ func (w *TraceWriter) Run() {
 	}
 }
 
+func (w *TraceWriter) runSync() {
+	defer close(w.stop)
+	defer close(w.flushChan)
+	for {
+		select {
+		case pkg := <-w.In:
+			w.addSpans(pkg)
+		case notify := <-w.flushChan:
+			w.drainAndFlush()
+			WaitForSenders(w.senders)
```

Comment:
This is a lot better in my opinion. It still feels a little bit out of place however. It's like a sudden switch of context between two things. I wish this code was closer to where we actually use `w.senders`, but that can get really complex. Regardless, I think we've come a long way and I hope you're also happy with the outcome.

No action to take here, just rambling and being open to any ideas/suggestions.

Reply:
We could potentially add an argument to drainAndFlush, eg `func drainAndFlush(waitForSenders bool)`? It's still one step separated from the flush call where we interact with the senders.

Reply:
Your idea lead me to another (point being to have all the senders logic together). What if we changed `sendPayloads` to:
```go
func sendPayloads(senders []*sender, p *payload, sync bool) {
```
And then call it as `sendPayloads(w.senders, p, w.syncMode)` ?

Would there be any risks with this? I think the premise is to wait for each flush in sync mode anyway. If you see no problems with this and like it too, perhaps we should just absorb the `WaitForSenders` logic into `sendPayloads` then.

Reply:
I've added a syncMode parameter to the sendPayloads function, and moved the sync logic in there.